File: server/services/branch_cleanup.py
"""
Branch Cleanup Service
======================
Cleans up remote feature branches on server startup.
"""

import asyncio
import logging
import subprocess
from pathlib import Path

logger = logging.getLogger(__name__)

PROTECTED_BRANCHES = {"main", "master", "beads-sync", "HEAD"}


async def cleanup_remote_branches_for_project(project_name: str, git_url: str, local_path: Path) -> int:
    """
    Delete all remote feature branches for a project.
    Uses the local clone at ~/.zerocoder/beads-sync/{project} or project path.

    Returns number of branches deleted.
    """
    from registry import get_beads_sync_dir

    # Use beads-sync clone path if available, otherwise project path
    work_dir = get_beads_sync_dir() / project_name
    if not work_dir.exists():
        work_dir = local_path
    if not work_dir.exists():
        logger.warning(f"No local clone found for {project_name}")
        return 0

    try:
        # Fetch to get current remote state
        subprocess.run(
            ["git", "-C", str(work_dir), "fetch", "--prune", "origin"],
            capture_output=True, timeout=60
        )

        # List remote branches
        result = subprocess.run(
            ["git", "-C", str(work_dir), "branch", "-r", "--format=%(refname:short)"],
            capture_output=True, text=True, timeout=30
        )
        if result.returncode != 0:
            return 0

        # Find branches to delete
        branches_to_delete = []
        for line in result.stdout.strip().split('\n'):
            if not line:
                continue
            branch = line.replace("origin/", "").strip()
            if branch and branch not in PROTECTED_BRANCHES:
                branches_to_delete.append(branch)

        if not branches_to_delete:
            return 0

        # Delete from remote
        deleted = 0
        for branch in branches_to_delete:
            result = subprocess.run(
                ["git", "-C", str(work_dir), "push", "origin", "--delete", branch],
                capture_output=True, text=True, timeout=30
            )
            if result.returncode == 0:
                deleted += 1
                logger.info(f"Deleted remote branch {branch} from {project_name}")

        return deleted

    except Exception as e:
        logger.warning(f"Error cleaning branches for {project_name}: {e}")
        return 0


async def cleanup_all_remote_branches() -> dict[str, int]:
    """
    Clean up remote feature branches for all registered projects.
    Called on server startup.

    Returns dict of project_name -> branches_deleted.
    """
    from registry import list_registered_projects, get_project_path, get_project_git_url

    results = {}
    projects = list_registered_projects()

    if not projects:
        return results

    logger.info(f"Cleaning up remote feature branches for {len(projects)} projects...")

    for name in projects:
        git_url = get_project_git_url(name)
        local_path = get_project_path(name)

        if not git_url or not local_path:
            continue

        deleted = await cleanup_remote_branches_for_project(name, git_url, local_path)
        if deleted > 0:
            results[name] = deleted

    if results:
        total = sum(results.values())
        logger.info(f"Cleaned up {total} remote feature branches across {len(results)} projects")

    return results

File: packages/server/src/services/branch-cleanup.ts
/**
 * Branch Cleanup Service
 * ======================
 *
 * Cleans up remote feature branches on server startup.
 */

import { exec } from 'child_process';
import { promisify } from 'util';
import { existsSync } from 'fs';
import { listRegisteredProjects, getProjectPath, getProjectGitUrl } from '../db/crud.js';

const execAsync = promisify(exec);

const FEATURE_BRANCH_PREFIX = 'feature/';

/**
 * Execute git command in project directory.
 */
async function runGit(workDir: string, args: string[], timeout = 30000): Promise<{ stdout: string; stderr: string }> {
  const { stdout, stderr } = await execAsync(`git ${args.join(' ')}`, {
    cwd: workDir,
    timeout,
  });
  return { stdout, stderr };
}

/**
 * Delete all remote feature branches for a project.
 * Uses the local project path.
 *
 * Returns number of branches deleted.
 */
export async function cleanupRemoteBranchesForProject(projectName: string, localPath: string): Promise<number> {
  if (!existsSync(localPath)) {
    console.warn(`No local clone found for ${projectName}`);
    return 0;
  }

  try {
    // Fetch to get current remote state
    await runGit(localPath, ['fetch', '--prune', 'origin'], 60000);

    // List remote branches
    const { stdout } = await runGit(localPath, ['branch', '-r', '--format=%(refname:short)'], 30000);

    // Find branches to delete
    const branchesToDelete: string[] = [];
    for (const line of stdout.trim().split('\n')) {
      if (!line) {
        continue;
      }
      const branch = line.replace('origin/', '').trim();
      if (branch && branch.startsWith(FEATURE_BRANCH_PREFIX)) {
        branchesToDelete.push(branch);
      }
    }

    if (branchesToDelete.length === 0) {
      return 0;
    }

    // Delete from remote
    let deleted = 0;
    for (const branch of branchesToDelete) {
      try {
        await runGit(localPath, ['push', 'origin', '--delete', branch], 30000);
        deleted++;
        console.info(`Deleted remote branch ${branch} from ${projectName}`);
      } catch (error) {
        console.warn(`Failed to delete branch ${branch} from ${projectName}:`, error);
      }
    }

    return deleted;
  } catch (error) {
    console.warn(`Error cleaning branches for ${projectName}:`, error);
    return 0;
  }
}

/**
 * Clean up remote feature branches for all registered projects.
 * Called on server startup.
 *
 * Returns dict of project_name -> branches_deleted.
 */
export async function cleanupAllRemoteBranches(): Promise<Record<string, number>> {
  const results: Record<string, number> = {};
  const projects = listRegisteredProjects();

  const projectNames = Object.keys(projects);
  if (projectNames.length === 0) {
    return results;
  }

  console.info(`Cleaning up remote feature branches for ${projectNames.length} projects...`);

  for (const name of projectNames) {
    const gitUrl = getProjectGitUrl(name);
    const localPath = getProjectPath(name);

    if (!gitUrl || !localPath) {
      continue;
    }

    const deleted = await cleanupRemoteBranchesForProject(name, localPath);
    if (deleted > 0) {
      results[name] = deleted;
    }
  }

  if (Object.keys(results).length > 0) {
    const total = Object.values(results).reduce((sum, count) => sum + count, 0);
    const projectCount = Object.keys(results).length;
    console.info(`Cleaned up ${total} remote feature branches across ${projectCount} projects`);
  }

  return results;
}
